### ISSUE TITLE
Cannot connect GSAP Enhancer to an element with an existing string ref

### ISSUE THREAD

**Issue by andreasnaslund** [azazdeaz/react-gsap-enhancer#3]
I can't get the react-gsap-enhancer to work properly. 

After I added:
import GSAP from 'react-gsap-enhancer';
...
and: 
@GSAP()
export default class MyComponent extends React.Component {
..

I get this error message:
{"message":"Cannot connect GSAP Enhancer to an element with an existing string ref. Please convert it to use a callback ref instead, or wrap it into a <span> or <div>. Read more: https://facebook.github.io/react/docs/more-about-refs.html#the-ref-callback-attribute"}

Any thoughts?

Thanks!

**Comment by azazdeaz** [comment]
Hi @andreasnaslund! 

Ref attributes can be strings or functions. If it is a function (callback ref) React will call it with the component when it's mounted. This enhancer using this callback refs to get the DOM of every element. If the element already have a callback ref it's going to be wrapped so nothing will change but it's can't be done with string refs.

**TLDR;** So the string refs have to be migrated to callback refs if the enhanced components.
Exaple: change ```<div ref='foo'/>``` to ```<div ref={comp => this.foo = comp}/>```

I copied this error message from [React DND](https://github.com/gaearon/react-dnd/blob/c8792616ac9c191001c2a94bfe06752c7f95f32a/src/utils/cloneWithRef.js#L8) but it's could be more helpfull. I will change it.

Hope this resolves your problem, thanks for trying react-gsap-enhancer out and making an issue!

[More About Refs](https://facebook.github.io/react/docs/more-about-refs.html#the-ref-callback-attribute)

**Comment by azazdeaz** [comment]
Hey @andreasnaslund, did this resolve your problem or you still have some trouble at using react-gsap-enhancer?

**Comment by micros123** [comment]
Hi, sorry to bump this issue, but I suspect a common gotcha with this solution. When setting the `ref` like that on a Component, you get the React class, not the DOM element. So you cannot animate it directly.
I'm now using `ReactDOM.findDOMNode(this.foo)`, which works fine, but it's not highly recommended because it breaks encapsulation. (It says so on the page you linked in the bottom of your post)
Anyway, for people looking for a quick fix:
`<MyComponent ref={ comp => this.foo = ReactDOM.findDOMNode(comp) }/>`

**Comment by azazdeaz** [comment]
Yes, this is true if you referencing a "composite component" (not like div, span, etc.).
Thanks for mentioning this. I updated my example above to be more general.

**Comment by jschertz** [comment]
If you're looking for ways to solve this issue, I present the following as _a_ solution, likely not _the_ solution for everyone:

```javascript
//At the top of the component, I declare a string variable for my ref
//Note: this is outside of my React component, 
//but inside the component file (usually above the component declaration
let greeting = 'greeting';
...

//Let's add the ref to our element. I write the callback function ref as such
<span 
  className="hello-world"
  ref={(node) => { greeting = node }}>
  Hello World!
</span>

...

//Now when I need to call upon my element with ref `greeting`, I call it like so:
let el = findDOMNode(this.refs.greeting);
console.log(el)
```
